File: java/leetcode/src/main/java/com/leetcode/graph/Max_Area_of_Island_695.java
package com.leetcode.graph;

public class Max_Area_of_Island_695 {
    int m = 0;
    int n = 0;
    int area = 0;
    int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public int maxAreaOfIsland(int[][] grid) {
        m = grid.length;
        n = grid[0].length;
        boolean[][] visited = new boolean[m][n];
        int max = 0;
        for (int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(!visited[i][j] && grid[i][j] == 1) {
                    dfs(grid, i, j, visited);
                    if (area > max) {
                        max = area;
                    }
                    area = 0;
                }
            }
        }
        return max;

    }

    void dfs(int[][] grid, int i, int j, boolean[][] visited) {
        visited[i][j] = true;
        area++;
        for(int d = 0; d <4; d++) {
            int[] dir = dirs[d];
            int row = i + dir[0];
            int col = j + dir[1];
            if (row >= 0 && row < m && col >= 0 && col < n && !visited[row][col] && grid[row][col] == 1) {
                dfs(grid, row, col, visited);
            }
        }

    }
}

File: js/algrithm/Graph/Max_Area_of_Island_695.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxAreaOfIsland = function(grid) {
    const dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    const m = grid.length
    const n = grid[0].length
    const visited = []
    for (let i = 0; i < m; i++) {
        visited.push(Array(n).fill(false))
    }

    const dfs = function(grid, i, j, visited) {
        visited[i][j] = true
        area++
        for (let dir of dirs) {
            let nr = i + dir[0]
            let nc = j + dir[1]
            if (nr >=0 && nr < m && nc >=0 && nc < n && !visited[nr][nc] && grid[nr][nc] === 1) {
                dfs(grid, nr, nc, visited)
            }
        }
    }

    let max = 0
    let area = 0
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (!visited[i][j] && grid[i][j] === 1) {
                dfs(grid, i, j, visited)
                if (area > max) {
                    max = area
                }
                area = 0
            }
        }
    }

    return max
};